service_capacity_modeling/tools/auto_shape.py

```diff
@@ -0,0 +1,166 @@
+import argparse
+import json
+import sys
+from collections import OrderedDict
+from fractions import Fraction
+from typing import Optional
+from typing import Sequence
+
+from service_capacity_modeling.hardware import shapes
+from service_capacity_modeling.interface import Drive
+from service_capacity_modeling.interface import ExcludeUnsetModel
+from service_capacity_modeling.interface import Instance
+from service_capacity_modeling.interface import InstancePricing
+
+
+class InstanceWithPricing(ExcludeUnsetModel):
+    name: str
+    instance: Instance
+    pricing: InstancePricing
+
+
+def pull_family(
+    instance_filter: str = "m7a", region: str = "us-east-1"
+) -> Sequence[InstanceWithPricing]:
+    # pylint: disable=unused-argument
+    """TODO: Talk to the AWS API
+
+    This will emulate
+
+    aws --region us-east-1 \
+      ec2 describe-instance-types \
+      --filters 'Name=instance-type,Values=m7a*'
+
+    And then convert from base 10 for memory and disk to base 2 and generate the
+    shape from that.
+
+    Returns a list of (name, instance, pricing). Then we can compose to writing
+    json files.
+    """
+    return []
+
+
+def reshape_family(
+    initial_shape: str, reference_family: str = "m5", region: str = "us-east-1"
+) -> Sequence[InstanceWithPricing]:
+    """Takes a single instance as a template and generates the other shapes
+
+    Useful for filling in details missing from automated pull such as IOPS
+    """
+    shape = shapes.hardware.regions[region].instances[initial_shape]
+    family, normalized_size = shape.family, shape.normalized_size
+
+    normalized_sizes = [
+        i.normalized_size
+        for i in shapes.hardware.regions[region].instances.values()
+        if i.family == reference_family
+    ]
+
+    def name(input_size: Fraction) -> str:
+        if input_size.denominator == 1 and input_size.numerator != 1:
+            size = str(input_size.numerator) + "xlarge"
+        else:
+            size = {
+                Fraction(1, 4): "medium",
+                Fraction(1, 2): "large",
+                Fraction(1, 1): "xlarge",
+            }[input_size]
```

Comment:
Should we also handle the `.metal-NN` sizes? I'd forgotten this! Low priority.

Reply:
Yeah metals are ... interesting ... let's take those separately